docs/arch/002-store-responses-in-query-string.md

```diff
@@ -0,0 +1,61 @@
+
+1. It is difficult to generate URLs with pre-populate responses, for example `/flow-name/y/response1/response2/<pre-populated-response>`. Unless the response is for the first question in a flow, it requires pre-filling responses for all former questions which are unlikely to be known. 
+
+1. Changes to a flow can cause issues with previously cached pages. As URLs are not deterministic for a particular page, changes to a flow can alter which page is displayed. For example, a given flow displays an outcome page for the path `/x/y/z`. A question is removed and the flow now shows an outcome page for the path `/x/y`. The cache would continue to return the old question and users would then be redirected to `/x/y/z`, which is no longer a valid path and could result in an error if it was not also cached.
+
+1. We lose existing responses if the user visits an earlier question. To view a previous or earlier question, responses need to be removed from the URL. This means that it is impossible to change a previous response without re-answering all subsequent questions. This can be a particular burden for users if the flow is long to complete, requires complex input (e.g. dates) or information that isn't memorised.
+
+Additionally, the implementation of storing user responses in a session required the creation of a seperate controller `SessionAnswersController`, separate logic for node resolution and updating the flow. This adds extra burden when maintaining or improving the codebase, as new code may need to accommodate both response storage methods.
+
+## Proposal
+
+Add the ability to store user responses using the URL query string. This will become the default method for storing responses and replace use of the URL path. The ability to store responses in a session cookie will remain and require flow to explicitly enable it within configuration (current behaviour).
+
+To implement this change, we could leverage the existing code (controller, node resolution and flow updates) written for storing responses in a session. This is because the underlying data structure storing the responses is the same i.e. a hash keyed by node name. The only additional logic required is to retrieve and update the parameters in the query string instead from the session.
+
+Example:
+- `flow-name/y` => `flow-name/s/question1`
+- `flow-name/y/response1` => `flow-name/s/question2?question1=response1`
+- `flow-name/y/response1/response2` => `flow-name/s/results?question1=response1&question2=question2`
+
+Suggested steps for implementation:
+
+1. Rename `SessionAnswersController` to `FlowController` (as will be used to handle all flows)
```

Comment:
Start page rendering behaviour is really confusing at the moment. The current behaviour is that:

- In our production, staging, integration environments, start pages are never rendered by Smart Answers and always Frontend
- In Heroku and developer environments, start pages are rendered by Smart Answers and by the SmartAnswersController, not the SessionAnswersController. We actually can't distinguish a session based vs path based smart answer at this point (`GET /flow_name`).

It's confusing because we [have code to render a start page in SessionAnswersController](https://github.com/alphagov/smart-answers/blob/b78e022315a6ce28108301638b09da6d4f1f8bc3/app/controllers/session_answers_controller.rb#L80), but it's never used. So collating controllers will clear that up. 

I'd really like to separate the start page rendering from the [controller method that renders the flow](https://github.com/alphagov/smart-answers/blob/b78e022315a6ce28108301638b09da6d4f1f8bc3/app/controllers/session_answers_controller.rb#L9). At the very least have a separate controller method, so we don't have to worry about non-production logic.

Reply:
Going back to your original question - 

I think our nomenclature is all over the place. The noun "Smart Answer" isn't well defined - do we mean the application itself? a flow? a flow + start page? If it is just flow, then why not just call it a flow?

It doesn't help that we don't even stick to the definition of a flow. We [say it's a collection of question and outcome nodes ](https://github.com/alphagov/smart-answers#nomenclature), however [`start_node`](https://github.com/alphagov/smart-answers/blob/b78e022315a6ce28108301638b09da6d4f1f8bc3/app/presenters/flow_presenter.rb#L92) is a thing within the FlowPresenter.  Also we use [`smart_answer`](https://github.com/alphagov/smart-answers/blob/b78e022315a6ce28108301638b09da6d4f1f8bc3/app/controllers/smart_answers_controller.rb#L53) as variable name, when actually its just a Flow object 😭. 

There's definitely work to be done to clear all of this up. My feeling is that "FlowController" is a good first step, as at least generalises the name from `SessionAnswersController` and prevents naming conflict with existing `SmartAnswersController` whilst we migrate. Possibly something to come back to later?

Reply:
Ha, my brain 🤯  you've opened my eyes to the cans of worms - somewhat amusing given there are so few actions.

So with the move to FlowController would that handle questions, responses and visualisation? and SmartAnswerController would cover the index and start pages?

Reply:
Ahah yup!

So we've currently got the following:

SmartAnswersController
- index (Index page)
- show (Start, questions and outcomes pages / Redirect to next question or outcome )
- visualise (Visualisation page)

SessionAnswersController
- start (Redirect to the first flow question)
- show (~Start~, question and outcome pages)
- update (Redirect to next question or outcome)
- destroy (Clear responses and redirect to start page)

Change to:

FlowController
- start (Redirect to the first flow question)
- show (Question and outcome pages)
- update (Redirect to next question or outcome)
- destroy (Clear responses and redirect to start page)

SmartAnswerController
- index (Index page)
- show (Start page)
- visualise (Visualisation page)

This seems like a good first step for what's in scope for the ADR. There is probably more we can do to tidy this up, for example improve controller names or method names, I'd like to defer that for now. Let me know if that makes sense, and if you agree I'll add this into the proposal.

Reply:
Thanks. Would visualise make more sense on FlowController since it's visualising a flow? Besides that controllers sound good. I think the action names on FlowController could use some work, how would you like to handle deferring that chat?

Reply:
So I think moving `visualise` is also out of scope for this ADR, because SmartAnswerController is going to still exist and it's not necessary for implementing storing responses in query strings. Probably also is related to discussion about improving controller and controller method names/organisation. 

(I've wondered if we want to group the index, start and visualise pages together as they are all thing strictly made for development and not user facing things. But I know that goes against typical conventions. Anyways part of the separate discussion definitely worth having.)

Shall we make a GitHub issue for this? Or this something we'd want to put the Tech Debt board?

Reply:
Yeah since visualise is publicly available I tend to group it with the flow stuff and only consider the start pages and index as the dev views. We do also have the concept of a DevelopmentController in other frontend apps, well at least there is one in EmailAlertFrontend.

A GitHub issue sounds good. I don't think we've got too much of a debt, just something a bit confusing that can be bikeshedded a bit.

Reply:
Started an issue #5225 - feel free to add anything I've missed.